python/source/state_representation/bind_parameters.cpp

```diff
@@ -0,0 +1,193 @@
+#include "state_representation_bindings.h"
+
+#include <state_representation/State.hpp>
+#include <state_representation/parameters/ParameterInterface.hpp>
+#include <state_representation/parameters/Parameter.hpp>
+#include <state_representation/space/cartesian/CartesianPose.hpp>
+#include <state_representation/robot/JointPositions.hpp>
+
+struct ParameterValues {
+  double double_value;
+  std::vector<double> double_array_value;
+  bool bool_value;
+  std::vector<bool> bool_array_value;
+  std::string string_value;
+  std::vector<std::string> string_array_value;
+  CartesianState cartesian_state;
+  CartesianPose cartesian_pose;
+  JointState joint_state;
+  JointPositions joint_positions;
+  Eigen::MatrixXd matrix_value;
+  Eigen::VectorXd vector_value;
+};
+
+class ParameterContainer : public ParameterInterface {
+public:
+  ParameterContainer(const std::string& name, const StateType& type) : ParameterInterface(type, name) {}
+
+  explicit ParameterContainer(const ParameterContainer& parameter) : ParameterInterface(parameter.get_type(), parameter.get_name()) {}
+
+  ParameterValues values;
+};
+
+void parameter_interface(py::module_& m) {
+  py::class_<ParameterInterface, State> c(m, "ParameterInterface");
+
+  c.def(py::init<const StateType&, const std::string&>(), "Constructor with parameter name and type of the parameter", "type"_a, "name"_a);
+  c.def(py::init<const ParameterInterface&>(), "Copy constructor from another ParameterInterface", "parameter"_a);
+}
+
+void parameter(py::module_& m) {
+  py::class_<ParameterContainer, ParameterInterface> c(m, "Parameter");
+
+  c.def(py::init<const std::string&, const StateType&>(), "Constructor of a parameter with name and type", "name"_a, "type"_a);
```

Comment:
I would like to see another signature for the constructor here which takes name, value and type in one go, without needing to set_value afterwards.

I suggest you actually define the set_value (and get_value) methods in the ParameterContainer class directly, rather than using a lambda function. Then you can add an appropriate constructor.

```
public:
  ParameterContainer(const std::string& name, const py::object& value, const StateType& type) 
      : ParameterInterface(type, name) {
    set_value(value);
  }
private:
  void set_value(const py::object& value) {
    switch(get_type()) {
      ...
      values.X = value.cast<X>()
      ...
    }
  }
```

The other reason I would prefer to have the set and get logic in the class is because I will have to re-use this class when I do the clproto encoding / decoding binding.